### ISSUE TITLE
cfg: not(...)-led comma list swallows later test operand (false negative)

### ISSUE THREAD

**Issue by dekobon** [dekobon/big-code-analysis#763]
## Summary

`cfg_predicate_marks_test` in `src/cfg_predicate.rs` mis-handles a
top-level comma list whose **first** operand is `not(...)` and whose
**last** operand ends in a closing paren. The `not(...)`
short-circuit fires on the *entire* list instead of just the first
operand, silently discarding the remaining operands — including a
`test` that should mark the item test-only. The result is a
**false negative**: such items are not recognized as test-only and
will not be stripped under `exclude_tests`.

## Location

`src/cfg_predicate.rs`, `cfg_predicate_marks_test`, lines 103-108:

```rust
if let Some(rest) = trimmed.strip_prefix("not").map(str::trim_start)
    && rest.starts_with('(')
    && rest.ends_with(')')
{
    continue;
}
```

The `not`-detection runs *before* the function splits `trimmed` into
top-level comma operands (the bare-comma-list branch at lines
127-129). When `trimmed` is a multi-operand list like
`not(foo), all(test)`, `strip_prefix("not")` yields `(foo), all(test)`,
which both `starts_with('(')` and `ends_with(')')`, so the guard
treats the whole list as a single `not(...)` operand and `continue`s
— dropping `all(test)` entirely.

## Reproduction

Verified against an extracted standalone copy of the module's logic:

```
cfg(not(foo),all(test))   = false   (expected true)
cfg(not(unix),any(test))  = false   (expected true)
cfg(all(not(foo),all(test))) = true (correct — bug does not fire when wrapped)
cfg(not(foo),test)        = true    (correct — list does not end in `)`)
```

Semantics: `cfg(not(foo), all(test))` is treated by this module as
`all(not(foo), all(test))` (the documented bare-comma-list = `all`
rule, lines 123-129). That predicate holds only when `foo` is absent
**and** `test` is set, so the item is test-only and must return
`true`. It returns `false`.

The asymmetry is the tell: `cfg(not(foo), test)` returns the correct
`true` purely because the list happens not to end in `)`. The
correctness of the result depends on the last operand's final
character, which is clearly wrong.

## Impact

- Test-only items annotated with a `not(...)`-led comma list ending
  in a parenthesized predicate are misclassified as production code
  and counted toward production metrics under `exclude_tests`.
- Severity is low in practice: the trigger requires the relatively
  unusual top-level form `cfg(not(X), <paren-predicate>)`. The
  equivalent wrapped form `cfg(all(not(X), all(test)))` is handled
  correctly because the outer `all(...)` splits the operands before
  the `not` guard sees them. Still, the module deliberately supports
  bare comma lists, so this is a genuine logic error in a documented
  code path, not merely undefined behaviour on malformed input.

## Suggested fix

Split into top-level operands *before* classifying, so each operand
is matched in isolation. Concretely, move the bare-comma-list branch
(lines 123-129) ahead of the `not`/`all`/`any` single-operand
classification: if `trimmed` has a top-level comma, push its operands
and `continue`; only when it is a single operand do the `not` /
`all` / `any` prefix checks apply. That makes the `ends_with(')')`
check operate on a true single operand, eliminating the
over-consumption.

## Regression test

Add to the `tests` module:

```rust
assert!(
    attribute_marks_test("cfg(not(foo), all(test))"),
    "not(...) first operand must not swallow a later test operand"
);
assert!(
    attribute_marks_test("cfg(not(unix), any(test))"),
    "not(...) short-circuit must apply only to its own operand"
);
```

Verify via revert: both assertions must fail against the current
lines 103-108 and pass after the reordering.

---

## Resolution

Fixed on `fix/issue-763` (commit 830eb662). The bare-comma-list split
now runs before the single-operand not/all/any classification in
`cfg_predicate_marks_test`, so each top-level operand is matched in
isolation. Paren-depth handling in `cfg_split_top_level_args` keeps
`not(foo, bar)` / `all(test, unix)` (commas inside a predicate's own
parens) as single operands. Regression test
`rust_attr_test_not_led_comma_list_keeps_later_test_operand` added and
verified via test-via-revert.

**Comment by dekobon** [comment]
### Fix plan

**Root cause.** `cfg_predicate_marks_test` classifies `trimmed` with
single-operand prefix tests (`not(`, `all(`, `any(`) before it splits a
top-level comma list into operands. A `not(...)`-led list that ends in
`)` therefore matches the `not` short-circuit as if it were one
operand, dropping every operand after the first.

**Change (src/cfg_predicate.rs, `cfg_predicate_marks_test`).**
Reorder the per-operand classification so multi-operand lists are split
first:

1. At the top of the loop body, after the `trimmed == "test"` check,
   test `cfg_split_top_level_args(trimmed).nth(1).is_some()`; if so,
   `stack.extend(cfg_split_top_level_args(trimmed))` and `continue`.
2. Only after that fallthrough (the operand is now guaranteed single)
   apply the `not(...)` short-circuit and the `all`/`any` descent.

This guarantees `ends_with(')')` in the `not` guard only ever sees a
single operand, so it can no longer span sibling operands.

**Why not just tighten the `not` guard.** Checking that the paren after
`not` is *balanced and closes at end of string* would also work, but
splitting-first is simpler, matches how `all`/`any` already recurse,
and removes the duplicated comma-split call at the end of the loop.

**Verification.**
- Add the two regression assertions from the issue body to the
  `tests` module.
- Test-via-revert: confirm both fail against current lines 103-108,
  pass after the reorder (per `.claude/rules/testing.md`).
- Re-run the existing `cfg_predicate` tests — especially
  `rust_attr_test_skips_not_test_and_feature_named_test` and the
  deeply-nested overflow test — to confirm no semantic regression for
  the single-operand `not(...)` and wrapped `all(not(...))` paths.
- `cargo test -p big-code-analysis cfg_predicate`.

**Scope.** Pure-logic, single-function change; no public API impact
(`attribute_marks_test` signature unchanged). No snapshot churn
expected (affects only `exclude_tests` classification of an unusual
attribute form not present in fixtures). Behaviour-only fix; confirm
no integration-snapshot drift before declaring done.

**Comment by dekobon** [comment]
Fixed on `fix/issue-763` (commit 830eb662).

**Root cause:** `cfg_predicate_marks_test` ran the single-operand
`not(...)` short-circuit *before* splitting `trimmed` into top-level
comma operands. For a list like `not(foo), all(test)`,
`strip_prefix("not")` yields `(foo), all(test)`, which both
`starts_with('(')` and `ends_with(')')`, so the guard treated the whole
list as one `not(...)` operand and `continue`d — discarding `all(test)`.
The bug surfaced only when the list's *last* operand ended in `)`, hence
the reported asymmetry.

**Fix:** Moved the bare-comma-list split (the `cfg_split_top_level_args(trimmed).nth(1).is_some()`
branch) ahead of the single-operand not/all/any prefix classification.
Each operand is now classified in isolation. `cfg_split_top_level_args`
already tracks paren depth, so a comma nested inside a predicate's own
parens (`not(foo, bar)`, `all(test, unix)`) is a single operand and is
unaffected.

**Verification (now all correct):**
- `cfg(not(foo), all(test))` → true
- `cfg(not(unix), any(test))` → true
- `cfg(all(not(foo), all(test)))` → true (was already correct)
- `cfg(not(foo), test)` → true (was already correct)
- `cfg(not(test))` → false
- `cfg(not(foo, bar))` → false (paren-nested comma, single operand)
- `cfg(all(test, unix))` → true (paren-nested comma)

**Tests:** Added `rust_attr_test_not_led_comma_list_keeps_later_test_operand`
in `src/cfg_predicate.rs`, verified test-via-revert (the two new
asserts fail against the pre-fix operand ordering). Full lib suite:
2908 passed, 0 failed. clippy/fmt clean.

CHANGELOG entry to add under `## [Unreleased]` (not edited per task constraints):
`### Fixed` — cfg: a `not(...)`-led top-level comma list ending in `)`
no longer swallows later operands, so a trailing `test` correctly marks
the item test-only (#763).
